kyaml/fn/framework/processors.go

```diff
@@ -0,0 +1,402 @@
+// It uses the given apiVersion and kind to do a simple lookup in the map and
+// returns an error if no exact match is found.
+func (m GVKFilterMap) ProviderFor(apiVersion, kind string) (kio.Filter, error) {
+	if kind == "" {
+		return nil, errors.Errorf("kind is required")
+	}
+	if apiVersion == "" {
+		return nil, errors.Errorf("apiVersion is required")
+	}
+
+	var ok bool
+	var versionMap map[string]kio.Filter
+	if versionMap, ok = m[kind]; !ok {
+		return nil, errors.Errorf("kind %q is not supported", kind)
+	}
+
+	var p kio.Filter
+	if p, ok = versionMap[apiVersion]; !ok {
+		return nil, errors.Errorf("apiVersion %q is not supported for kind %q", apiVersion, kind)
+	}
+	return p, nil
+}
+
+// FilterProvider is implemented by types that provide a way to look up which Filter
+// should be used to process a ResourceList based on the ApiVersion and Kind of the
+// ResourceList.functionConfig in the input. FilterProviders are intended to be used
+// as part of VersionedAPIProcessor.
+type FilterProvider interface {
+	// ProviderFor returns the appropriate filter for the given APIVersion and Kind.
+	ProviderFor(apiVersion, kind string) (kio.Filter, error)
+}
+
+// VersionedAPIProcessor selects the appropriate kio.Filter based on the ApiVersion
+// and Kind of the ResourceList.functionConfig in the input.
+// It can be used to implement configuration function APIs that evolve over time,
+// or create processors that support multiple configuration APIs with a single entrypoint.
+// All provided Filters MUST be structs capable of receiving ResourceList.functionConfig data.
+// Provided Filters MAY implement Defaulter and Validator to have Default and Validate
+// respectively called between unmarshalling and filter execution.
+type VersionedAPIProcessor struct {
+	// FilterProvider resolves a kio.Filter for each supported API, based on its APIVersion and Kind.
+	// GVKFilterMap is a simple FilterProvider implementation for use here.
+	FilterProvider FilterProvider
+}
+
+// Process makes VersionedAPIProcessor implement the ResourceListProcessor interface.
+// It looks up the configuration object to use based on the ApiVersion and Kind of the
+// input ResourceList.functionConfig, loads ResourceList.functionConfig into that object,
+// invokes Validate and Default if supported, and finally invokes Filter.
+func (p *VersionedAPIProcessor) Process(rl *ResourceList) error {
+	api, err := p.FilterProvider.ProviderFor(extractGVK(rl.FunctionConfig))
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	if err := LoadFunctionConfig(rl.FunctionConfig, api); err != nil {
+		return errors.Wrap(err)
+	}
+	return errors.Wrap(rl.Filter(api))
+}
+
+// extractGVK returns the apiVersion and kind fields from the given RNodes if it contains
+// valid TypeMeta. It returns an empty string if a value is not found.
+func extractGVK(src *yaml.RNode) (apiVersion, kind string) {
+	if src == nil {
+		return
```

Comment:
Even though the language allows it, it's better to avoid naked returns since they can lead to unexpected behavior.

Here I'd do `return "", ""`

And below `return apiVersion, kind`